### DISCUSSION TITLE
Impersonation with two concurrent sessions

### DISCUSSION THREAD

**Discussion by wyozi** [supabase/supabase#10237] [Questions]
Hey!

I'm trying to implement impersonation feature (where the primary session changes to that of someone else, while the real admin user is still kept in another session cookie).

I figured that by using auth-helpers to have two session concurrently, this would be possible. I.e. something like this:
```ts
// in pages/api/impersonateUser.ts
const USER_TO_IMPERSONATE = "test@example.com";

async function handler(req: NextApiRequest, res: NextApiResponse) {
  const client = createServerSupabaseClient({ req, res });
  const realClient = createServerSupabaseClient(
    { req, res },
    {
      cookieOptions: {
        name: "supabase-real-auth-token",
        domain: undefined,
        path: "/",
        sameSite: "lax",
        secure: false,
        maxAge: 1000 * 60 * 60 * 24 * 365,
      },
    }
  );

  const { data } = await client.auth.getSession();
  await realClient.auth.setSession(data!.session!);

  const fakeLink = await supabaseServiceClient().auth.admin.generateLink({
    type: "magiclink",
    email: USER_TO_IMPERSONATE,
    options: {
      redirectTo: "http://localhost:4040",
    },
  });
  return res.status(307).redirect(fakeLink.data.properties!.action_link!);
}
```

This works good enough that it changes the primary session to that of the impersonated user and creates the "supabase-real-auth-token" local storage variable as well. However, the "supabase-real-auth-token" token becomes equal to the impersonated one.

Is having multiple supabase sessions simultaneously (with auth-helpers or otherwise) functionally possible at all?

**Comment by j4w8n** [comment]
@wyozi I have high level steps worked out for this. It would be a feature of the supabase client. I can't work on it until next week, and have no clue whether it would be implemented.

**Reply by wyozi** [reply]
I found the dual session approach a bit awkward anyway so ended up coming with an alternative approach that I documented here: https://catjam.fi/articles/supabase-admin-impersonation

**Reply by j4w8n** [reply]
Ok, thanks for the link!
